### ISSUE TITLE
Allow migration from Basic LB to Standard

### ISSUE THREAD

**Issue by rlevchenko** [Azure/AKS#1687]
### GA ETA: August 2025*
  *ETAs are estimations and are subject to change

[request] There are some features that rely on Standard LB only and aren't allowed on cluster with Basic Load Balancers. For example, we can't configure authorized IP ranges to limit access to the API server for now. If we don't have techinical limits, can you allow us to move from Basic LB to Standard LB without re-creating an entire cluster? Any "hidden" or even "unoffical" walkarounds?

**Comment by rlevchenko** [comment]
Any thoughts? Since we have this guide https://docs.microsoft.com/en-us/azure/load-balancer/upgrade-basicinternal-standard , would it be the problem to allow such migration in terms of the AKS?

**Comment by TomGeske** [comment]
@palma21

**Comment by jluk** [comment]
Ack on this ask, it is in a discussion on the AKS product team to address this pain point in migrating from BLB to SLB.

**Comment by palma21** [comment]
Re-opening so we may track on the backlog and provide updates. Let me know if you'd rather us create a new one

**Comment by songkranj** [comment]
A lot of clusters are waiting for this features.

**Comment by junkiebev** [comment]
Until this is developed, can y'all have the networking team at MS stop squeezing the capabilities of the basic load balancer down? It's punitive to earlier adopters and is causing the entire platform to be viewed as not stable.

**Comment by MattiasVahlberg** [comment]
@palma21 Do we have any update if this under development or not? We created an AKS earlier this year and did not get a prompt for basic vs standard so it defaulted to standard. We successfully created another nodepool that we've used and it worked as expected. However, we are now unable to deploy a new node pool due to basic load balancer. Is it possible to get an ETA on this feature?

**Comment by rlevchenko** [comment]
"Azure public IP addresses now support the ability to be upgraded from Basic to Standard SKU.  Additionally, any Basic Public Load Balancer can now be upgraded to a Standard Public Load Balancer, while retaining the same public IP address.  This is supported via PowerShell, CLI, templates, and API and available across all Azure regions" Is it also supported by AKS?

**Comment by astraverkhau** [comment]
Much needed for long-lived AKS that holds a lot of barely movable state.

**Comment by Habulinec** [comment]
Looking forward for any update on this as well. :)

**Comment by lissff** [comment]
@palma21  any ETA on the migration

**Comment by benydc** [comment]
@palma21 any update on the migration?

**Comment by Tbohunek** [comment]
This would be great.

**Comment by AsoTora** [comment]
Looking forward for this as well. Would be really nice!

**Comment by romanlum** [comment]
+1

**Comment by bmenas** [comment]
+1

**Comment by jessereinhart** [comment]
+1!

I'm in a weird catch 22 involving somehow having two node pools and a BLB. I can't add a node pool because I've got BLB, so my memory-heavy workloads are forced to run on non-optimized instance types.

**Comment by burtsimpson** [comment]
+1

Multiple pools would allow us to better designate targets in Azure DevOps pipelines.

**Comment by edberg** [comment]
@palma21 Any updates on the progress on this issue?

**Comment by rlevchenko** [comment]
Meanwhile , upgrade script for a standalone LB SKU is GA - https://azure.microsoft.com/en-us/updates/load-balancer-sku-upgrade-now-available-through-powershell-script-2/ I know, it's not the same and we need some actions from AKS team anyway.

**Comment by kiebrew** [comment]
+1

**Comment by justinkaufman** [comment]
+1

**Comment by ludwig-reekoh** [comment]
+1

**Comment by vutkin** [comment]
+1

**Comment by StefanNa3Shape** [comment]
+1

**Comment by Sjd-Risca** [comment]
+1

**Comment by iagomez** [comment]
+1

**Comment by cthulhu** [comment]
Any updates on this? I used the script and now LB is standard, but AKS still thinks that it is Basic. Is there a way to "tell" to  AKS that LB is Standard and we can add more nodepools?

**Comment by FlorentATo** [comment]
@cthulhu Does that mean you still see the banner "This cluster is using a basic load balancer. Having multiple node pools is not supported." in AKS > Nodepools, despite the loadbalancer being now SKU: Standard ?

Have you tried adding new nodepools to AKS using the CLI ([az aks nodepool add](https://learn.microsoft.com/en-us/cli/azure/aks/nodepool?view=azure-cli-latest#az-aks-nodepool-add)) ? Do you get any error message ?

**Comment by kevin-schmid** [comment]
I have a similar state as @cthulhu. I tried upgrading my cluster with `Start-AzBasicLoadBalancerUpgrade -ResourceGroupName management-cluster-nodes -BasicLoadBalancerName kubernetes` which results in a load balancer with standard sku

```
> Get-AzLoadBalancer -ResourceGroupName management-cluster-nodes

ResourceGroupName        Name       Location   ProvisioningState Sku Name
-----------------        ----       --------   ----------------- --------
management-cluster-nodes kubernetes westeurope Succeeded         Standard
```

but the network profile of the cluster says I still use basic:

```
> az aks show --resource-group management-cluster --name management-cluster
{
  "aadProfile": null,
  "addonProfiles": null,
  "agentPoolProfiles": [
    {
      ...
      "type": "VirtualMachineScaleSets",
      "upgradeSettings": {
        "maxSurge": null
      },
      "vmSize": "Standard_B2s",
      "workloadRuntime": null
    }
  ],
  ...
  "name": "management-cluster",
  "networkProfile": {
    "dnsServiceIp": "10.0.0.10",
    "dockerBridgeCidr": "172.17.0.1/16",
    "ipFamilies": [
      "IPv4"
    ],
    "loadBalancerProfile": null,
    "loadBalancerSku": "Basic",
    "natGatewayProfile": null,
    "networkMode": null,
    "networkPlugin": "kubenet",
    "networkPluginMode": null,
    "networkPolicy": null,
    "outboundType": "loadBalancer",
    "podCidr": "10.244.0.0/16",
    "podCidrs": [
      "10.244.0.0/16"
    ],
    "serviceCidr": "10.0.0.0/16",
    "serviceCidrs": [
      "10.0.0.0/16"
    ]
  },
  "nodeResourceGroup": "management-cluster-nodes",
  "oidcIssuerProfile": {
    "enabled": false,
    "issuerUrl": null
  },
  "podIdentityProfile": null,
  "powerState": {
    "code": "Running"
  },
  "privateFqdn": null,
  "privateLinkResources": null,
  "provisioningState": "Succeeded",
  "publicNetworkAccess": "Enabled",
  "resourceGroup": "management-cluster",
  "securityProfile": {
    "azureKeyVaultKms": null,
    "defender": null,
    "workloadIdentity": null
  },
  "servicePrincipalProfile": {
    "clientId": "msi"
  },
  "sku": {
    "name": "Basic",
    "tier": "Free"
  },
  ...
}
```

**Comment by rlevchenko** [comment]
In summary, the feature was added to the backlog almost 3 years ago; no activities since then (based on the backlog status)
Any updates? ETA? It'd be interesting to read about tech. issues you faced with, if any. Thanks.

**Comment by cthulhu** [comment]
We ended up with cluster recreation. Create a new cluster with -v1 suffix, migrate workloads, test, recreate original cluster, migrate workload, test.

**Comment by JonasJes** [comment]
Azure has started sending retirement warnings via email "Action required: Upgrade from Basic to Standard SKU public IP addresses in Azure by 30 September 2025"

This has not been resolved for almost 4 years now.
So what are our options, other than recreating our clusters?

**Comment by thpou** [comment]
We are a big company here that needs to transition to the standard LB before September 30 2025 like everyone else. We recently tried to migrate our test cluster and to our big surprise discovered it cannot be done without recreating the cluster as people here has done.  

Difference here is that we are a big global company that cannot allow for the downtime a recreation of the cluster will introduce. 

We have decided to postpone the task of transitioning from basic LB to standard LB another 6 months and we sincerely expect you to come up with at least some clarification on what to expect. Are you going to provide us with a painless migration approach or are we to prepare our customers for the inevitable downtime ?

**Comment by Tbohunek** [comment]
@thpou if that helps at all... I've built a 2nd Cluster in parallel, run the same App(s) there as form of HA. Then I flipped the FrontDoor backend from old Cluster to new Cluster without customers noticing. Same would work with App GW.

**Comment by orsharon7** [comment]
+1

**Comment by miblafi** [comment]
Any progress? We need to replace this on several AKS

**Comment by wdarko1** [comment]
We are currently working on a migration path from Basic Load Balancer to Standard. Will follow up with an expected target date for availability.

**Comment by wdarko1** [comment]
> Please guide here how we can upgrade AKS loadbalancer SKU from Basic to Standard

The existing migration method is detailed in the following doc - [Microsoft Learn - Upgrade Basic LB to Standard in Powershell](https://learn.microsoft.com/en-us/azure/load-balancer/upgrade-basic-standard-with-powershell)

As mentioned, we will be providing a simplified CLI command to perform this migration, which will be available early 2025.

**Comment by kevin-schmid** [comment]
> > Please guide here how we can upgrade AKS loadbalancer SKU from Basic to Standard
> 
> The existing migration method is detailed in the following doc - [Microsoft Learn - Upgrade Basic LB to Standard in Powershell](https://learn.microsoft.com/en-us/azure/load-balancer/upgrade-basic-standard-with-powershell)
> 
> As mentioned, we will be providing a simplified CLI command to perform this migration, which will be available early 2025. 
> 

This does not work for AKS, at least when I tried it as mentioned [in here before.](https://github.com/Azure/AKS/issues/1687#issuecomment-1578646495) The network profile in AKS does not change and the restrictions of a Basic LB still apply to the cluster (i.e. single type node pool).

**Comment by jansauer** [comment]
@wdarko1 can you please confirm that the "simplified CLI command" you are working on also extends the migration to work on Basic Load Balancers for Azure Kubernetes Services (AKS) clusters?

The problem described in this issue is not that it is to complicated but that the document you linked to explicitly excludes AKS and does not work there.

**Comment by larsf96** [comment]
@wdarko1 Will this also work for changing Standard Load Balancers back to Basic?

**Comment by wdarko1** [comment]
Correction - For AKS clusters, the current method to migrate from the Basic Load balancer to Standard is to spin up a new cluster and manually migrate, as the earlier linked method does not include AKS. 

The upcoming CLI command will be the simplified migration path specifically for AKS clusters. This is scheduled to become available in early 2025. @kevin-schmid @jansauer 


@larsf96 This CLI command migration path will only move from the Basic Load Balancer to Standard load balancer, as the Basic load balancer will be discontinued in Sept 2025. Additionally new clusters will no longer allow the Basic load balancer in March 2025 [See official announcement](https://azure.microsoft.com/en-us/updates?id=azure-basic-load-balancer-will-be-retired-on-30-september-2025-upgrade-to-standard-load-balancer)

**Comment by neoGeneva** [comment]
Hey @wdarko1, just to clarify what the upcoming CLI command will actually does. Does it take an existing cluster and upgrade its load balancer, or does it automatically spin up a new cluster and migrate across to it?

**Comment by Johpie** [comment]
Hey @wdarko1, is there any update regarding this, since we need to decide if we are going to rebuild our AKS clusters or wait for this update.

**Comment by v-braun** [comment]
@wdarko1 This would be really awesome.
At least have the possibility that new LoadBalancers created will have the Standard tier.

Our scenario:
We cannot destroy / recreate the AKS, BUT it would be fine for us to destroy and create a new LB as long the new LB will have the Standard SKU.
Unfortunatley it is currently not possible to update the "default" LoadBalancer SKU.

We look forward to May before we migrate.

**Comment by stefano-salari** [comment]
Hello @wdarko1, any update on the release of the CLI script?
Like many other people I'm looking forward this tool for the migration of our obsolete AKS load balancers, and giving that the deadline is approaching, I'm a bit concerned about the time of its release.

Many thanks in advance!

**Comment by lturmel** [comment]
@wdarko1 It's way too late for my client to wait for that CLI to allow a load balancer upgrade for AKS.

I'm more than very disappointed that this CLI to allow such upgrade/migration hasn't been publish the date of the retirement announcement or at max 6 to 9 months before the retirement date.

For my client, we just create a new cluster and migrate everything on it... but this is a very costly migration.

**Comment by majooneez** [comment]
Is there any update on the date when this script will be available? It's almost end of May and not sure how to proceed, if I should start working on new clusters already or there will be automation script for that migration really available in reasonable time before retirement of Basic SKU.

**Comment by sata-sa** [comment]
Any updates on the AZ CLI command to perform this action?

**Comment by BettyMichel** [comment]
@wdarko1  Any update on this ?

**Comment by markuspaller** [comment]
Q2 is over in a few days, and we are also waiting for this feature urgently. Please provide a solution as fast as possible, as the deadline in coming closer and closer, and vacation season is coming up. 
Thanks

**Comment by wdarko1** [comment]
The feature is rolling out across regions with our next AKS release (v20250617) in Preview. To track when the feature reaches your region, visit [AKS Release Status](https://releases.aks.azure.com/)

Public documentation is available here: [Upgrading the Basic Load Balancer on AKS](https://learn.microsoft.com/en-us/azure/aks/upgrade-basic-load-balancer-on-aks).

Note: for any users who need to migrate clusters from Availability Sets also, there is a separate script that performs both migrations at once - see the documentation here: [Migration from Availability Sets on AKS](https://learn.microsoft.com/en-us/azure/aks/availability-sets-on-aks)

**Comment by ybdc-jl** [comment]
Hi @wdarko1 thank you for the update.
One issue I was facing during the migration and detailed trace was attached below:
```
❯  az aks update --name {cluster-name} --resource-group {resource-group} --load-balancer-sku=Standard

// skipped

cli.azure.cli.core.azclierror: Traceback (most recent call last):
  File "/usr/lib64/az/lib/python3.12/site-packages/knack/cli.py", line 233, in invoke
    cmd_result = self.invocation.execute(args)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/az/lib/python3.12/site-packages/azure/cli/core/commands/__init__.py", line 666, in execute
    raise ex
  File "/usr/lib64/az/lib/python3.12/site-packages/azure/cli/core/commands/__init__.py", line 734, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/az/lib/python3.12/site-packages/azure/cli/core/commands/__init__.py", line 703, in _run_job
    result = cmd_copy(params)
             ^^^^^^^^^^^^^^^^
  File "/usr/lib64/az/lib/python3.12/site-packages/azure/cli/core/commands/__init__.py", line 336, in __call__
    return self.handler(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/az/lib/python3.12/site-packages/azure/cli/core/commands/command_operation.py", line 120, in handler
    return op(**command_args)
           ^^^^^^^^^^^^^^^^^^
  File "/home/blurred/.azure/cliextensions/aks-preview/azext_aks_preview/custom.py", line 971, in aks_update
    return aks_update_decorator.update_mc(mc)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/az/lib/python3.12/site-packages/azure/cli/command_modules/acs/managed_cluster_decorator.py", line 9221, in update_mc
    return self.put_mc(mc)
           ^^^^^^^^^^^^^^^
  File "/home/blurred/.azure/cliextensions/aks-preview/azext_aks_preview/managed_cluster_decorator.py", line 5691, in put_mc
    cluster = sdk_no_wait(
              ^^^^^^^^^^^^
  File "/usr/lib64/az/lib/python3.12/site-packages/azure/cli/core/util.py", line 744, in sdk_no_wait
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/az/lib/python3.12/site-packages/azure/core/tracing/decorator.py", line 94, in wrapper_use_tracer
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/blurred/.azure/cliextensions/aks-preview/azext_aks_preview/vendored_sdks/azure_mgmt_preview_aks/v2025_05_02_preview/operations/_managed_clusters_operations.py", line 2094, in begin_create_or_update
    raw_result = self._create_or_update_initial(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/blurred/.azure/cliextensions/aks-preview/azext_aks_preview/vendored_sdks/azure_mgmt_preview_aks/v2025_05_02_preview/operations/_managed_clusters_operations.py", line 1960, in _create_or_update_initial
    raise HttpResponseError(response=response, error_format=ARMErrorFormat)
azure.core.exceptions.HttpResponseError: (OperationNotAllowed) Tenant change detected. Subscription 'blurred-c8d0-456a-ad3c-blurred' was moved from tenant: 'blurred-cd1a-4285-92fc-blurred' to 'blurred-13d9-4a35-965f-blurred'. Moving AKS clusters between tenants is currently unsupported. Please move the subscription back or delete and recreate this cluster.
Code: OperationNotAllowed
Message: Tenant change detected. Subscription 'blurred-c8d0-456a-ad3c-blurred' was moved from tenant: 'blurred-cd1a-4285-92fc-blurred' to 'blurred-13d9-4a35-965f-blurred'. Moving AKS clusters between tenants is currently unsupported. Please move the subscription back or delete and recreate this cluster.

cli.azure.cli.core.azclierror: (OperationNotAllowed) Tenant change detected. Subscription 'blurred-c8d0-456a-ad3c-blurred' was moved from tenant: 'blurred-cd1a-4285-92fc-blurred' to 'blurred-13d9-4a35-965f-blurred'. Moving AKS clusters between tenants is currently unsupported. Please move the subscription back or delete and recreate this cluster.
Code: OperationNotAllowed
Message: Tenant change detected. Subscription 'blurred-c8d0-456a-ad3c-blurred' was moved from tenant: 'blurred-cd1a-4285-92fc-blurred' to 'blurred-13d9-4a35-965f-blurred'. Moving AKS clusters between tenants is currently unsupported. Please move the subscription back or delete and recreate this cluster.
az_command_data_logger: (OperationNotAllowed) Tenant change detected. Subscription 'blurred-c8d0-456a-ad3c-blurred' was moved from tenant: 'blurred-cd1a-4285-92fc-blurred' to 'blurred-13d9-4a35-965f-blurred'. Moving AKS clusters between tenants is currently unsupported. Please move the subscription back or delete and recreate this cluster.
Code: OperationNotAllowed
Message: Tenant change detected. Subscription 'blurred-c8d0-456a-ad3c-blurred' was moved from tenant: 'blurred-cd1a-4285-92fc-blurred' to 'blurred-13d9-4a35-965f-blurred'. Moving AKS clusters between tenants is currently unsupported. Please move the subscription back or delete and recreate this cluster.
cli.knack.cli: Event: Cli.PostExecute [<function AzCliLogging.deinit_cmd_metadata_logging at 0x7fec4cee0720>]
az_command_data_logger: exit code: 1
````
It seems the command failed to recognize the current tenant (this particular AKS was moved from another tenant 5 month ago).
Could you please confirm if this is the expected behavior, or if it might be a bug? Any guidance or suggestions would be appreciated.

**Comment by wdarko1** [comment]
@ybdc-jl This is the expected behavior. As mentioned in the error message, tenant moves are not supported in AKS. Any `az aks update` command will likely receive a similar error code, unless moving to the original tenant. See [AKS FAQ documentation](https://learn.microsoft.com/azure/aks/faq#can-i-move-or-migrate-my-cluster-between-azure-tenants-).

`Message: Tenant change detected. Subscription 'blurred-c8d0-456a-ad3c-blurred' was moved from tenant: 'blurred-cd1a-4285-92fc-blurred' to 'blurred-13d9-4a35-965f-blurred'. Moving AKS clusters between tenants is currently unsupported. Please move the subscription back or delete and recreate this cluster.`

**Comment by JonasJes** [comment]
We migrated today, and it is now up and running with the standard load balancer.

Background:
Our cluster is 3 nodepools, 22 total Linux nodes with 352 pods

Notes:
- The migration script was running for about 20 minutes with downtime on the whole cluster
- After migration, we were still experiencing problems; some pods were running, some were not.
- Proxy to Kubernetes Dashboard was among the not working, so we rebooted the cluster, and after that, Kubernetes Dashboard was available again.
   - I am not sure if this was needed
- But the main problem was that the **outbound IP was changed**, so most of the applications were hitting firewalls, and that was why they were not working.

But the [documentation ](https://learn.microsoft.com/en-us/azure/aks/upgrade-basic-load-balancer-on-aks) says it should have been the same, but is that only true for the inbound
> This process will also migrate your Basic IP to a Standard IP, while keeping the IP addresses the same.

@wdarko1, You probably should look into this or change the documentation

Meanwhile, I would recommend running this after the migration to validate the IP
``` bash
# Get the outbound IP Resource ID
az aks show -g <myResourceGroup> -n <myAKSCluster> --query networkProfile.loadBalancerProfile.effectiveOutboundIPs[].id
# get the new IP for each IP Resource ID
az network public-ip show --ids <IPResourceID> --query ipAddress -o tsv
 ```

**Comment by wdarko1** [comment]
Hi @JonasJes thanks for the feedback - the documentation has been updated to clarify that the inbound IP addresses will remain the same while there will be new outbound IP addresses used.  See the [updated documentation](https://learn.microsoft.com/en-us/azure/aks/upgrade-basic-load-balancer-on-aks#before-you-begin) for more info.  We also updated the documentation to include the commands to validate the IP addresses.

**Comment by kevin-schmid** [comment]
We also managed to migrate today. We had the issue that 2 outboundRules existed (I guess due to my failed migration attempt with Start-AzBasicLoadBalancerUpgrade). `az network lb outbound-rule list` only listed 1 rule named "kubernetes" but cluster reconciliation mentioned as second rule named "aksOutboundRule". We had to delete the "kubernetes" outboundRule.

**Comment by Tomasz-Kluczkowski** [comment]
Does this upgrade both the public load balancer of an AKS cluster from Basic and internal load balancers too?

For one of my clusters I can see 2 load balancers...

<img width="748" height="110" alt="Image" src="https://github.com/user-attachments/assets/2e34d4e8-56fb-4d4a-bbad-fd90792e3e97" />

**Comment by palma21** [comment]
> Does this upgrade both the public load balancer of an AKS cluster from Basic and internal load balancers too?
> 
> For one of my clusters I can see 2 load balancers...
> 
> <img alt="Image" width="748" height="110" src="https://private-user-images.githubusercontent.com/26039401/480134591-2e34d4e8-56fb-4d4a-bbad-fd90792e3e97.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzEzNzQ1ODgsIm5iZiI6MTc3MTM3NDI4OCwicGF0aCI6Ii8yNjAzOTQwMS80ODAxMzQ1OTEtMmUzNGQ0ZTgtNTZmYi00ZDRhLWJiYWQtZmQ5MDc5MmUzZTk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjE4VDAwMjQ0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlYjg1YWJlNzQxMjRkNGRmOGRiZWU4NjVmMzIyMjFiMTVhMzdlY2NmYTNiNWZkODkxZThmODNlMjBkZjNiM2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.riFUCC_Ses8lpfa2bFqKreo1TxnvOmMxJAJY7psoxEU">

yes
